python/cudf/cudf/core/df_protocol.py

```diff
@@ -0,0 +1,729 @@
+            dtype = self.dtype
+        elif self.dtype[0] == _k.BOOL:
+            # convert bool to uint8 as dlpack does not support bool natively.
+            buffer = _CuDFBuffer(
+                cp.asarray(self._col.fillna(invalid).to_gpu_array(), dtype=cp.uint8),
+                allow_copy=self._allow_copy)
+            dtype = self.dtype
+        elif self.dtype[0] == _k.CATEGORICAL:
+            codes = self._col.cat.codes
+            buffer = _CuDFBuffer(
+                cp.asarray(codes.fillna(invalid)),
+                allow_copy=self._allow_copy)
+            dtype = self._dtype_from_cudfdtype(codes.dtype)
+        # elif self.dtype[0] == _k.STRING:
+        #     # Marshal the strings from a NumPy object array into a byte array
+        #     buf = self._col.to_numpy()
+        #     b = bytearray()
+
+        #     # TODO: this for-loop is slow; can be implemented in Cython/C/C++ later
+        #     for i in range(buf.size):
+        #         if type(buf[i]) == str:
+        #             b.extend(buf[i].encode(encoding="utf-8"))
+
+        #     # Convert the byte array to a Pandas "buffer" using a NumPy array as the backing store
+        #     buffer = _CuDFBuffer(np.frombuffer(b, dtype="uint8"))
+
+        #     # Define the dtype for the returned buffer
+        #     dtype = (_k.STRING, 8, "u", "=")  # note: currently only support native endianness
+        else:
+            raise NotImplementedError(f"Data type {self._col.dtype} not handled yet")
+
+        return buffer, dtype
+
+    def _unpackbits(self, myarray, bitorder="big"):
+    
+        bitorder_op = {"big": '(myarray[i / 8] >> (7 - i % 8)) & 1;', 
+                        "little": '(myarray[i / 8] >> (i % 8)) & 1;'}
+        operation = bitorder_op.get(bitorder, None)
+        if operation == None:
+            raise KeyError(f"bitorder must be either 'big' or 'little' not '{bitorder}'")
+        _unpackbits_kernel = _core.ElementwiseKernel(
+        'raw uint8 myarray', 'T unpacked',
+        'unpacked = '+ operation,
+        'unpackbits_kernel'
+        )
+
+        if myarray.dtype != cp.uint8:
+            raise TypeError('Expected an input array of unsigned byte data type')
+
+        unpacked = cp.ndarray((myarray.size * 8), dtype=cp.uint8)
+        return _unpackbits_kernel(myarray, unpacked)
+
+    def _get_validity_buffer(self) -> Tuple[_CuDFBuffer, Any]:
+        """
+        Return the buffer containing the mask values indicating missing data and
+        the buffer's associated dtype.
+
+        Raises RuntimeError if null representation is not a bit or byte mask.
+        """
+        
+        null, invalid = self.describe_null
+        if null == 3:
+            _k = _DtypeKind
+            bitmask = self._unpackbits(cp.asarray(self._col._column.mask), bitorder='little')[:len(self._col)]
+            buffer = _CuDFBuffer(bitmask)
+            dtype = (_k.UINT, 8, "C", "=")
+            return buffer, dtype
```

Comment:
As I learned today, there is a utility in cuDF to get the bitmask of a column as a byte mask, that we could use here instead:

```python

In [30]: df
Out[30]:
      a
0     1
1  <NA>
2     2
3  <NA>
4  <NA>
5     3

In [31]: df["a"]._column._get_mask_as_column()
Out[31]:
<cudf.core.column.numerical.NumericalColumn object at 0x7f7bc1049b90>
[
  true,
  false,
  true,
  false,
  false,
  true
]
dtype: bool
```

Reply:
Was looking for such function ! Very nice to know about that function. I can get rid of the `unpackbits` code.